### PR TITLE
Pcvl 598 a job group monitor

### PR DESCRIPTION
**Pull request by raksharuia**

### CODE REVIEW

File: perceval/runtime/job_group.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+            warnings.warn(UserWarning(f"{exc}"))
+
+        # validate read/write permissions for the directory
+        if not os.access(dir_path, os.W_OK) or not os.access(dir_path, os.R_OK):
+            warnings.warn(UserWarning(f"Cannot read or write in {dir_path}"))
+
+    def _create_job_group(self):
+        # Saves information for a new job group on disk using PersistentData(),
+        jgrp_dir_path = JobGroup._get_job_group_dir()
+        JobGroup._setup_job_group_dir(jgrp_dir_path)  # create directory and validat permissions
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:100]: create directory and validat**E** permissions

File: perceval/runtime/job_group.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        self._modify_job_dataset(job_info)
+
+    def _modify_job_dataset(self, updated_info: dict):
+        # modifies the recorded information when a new job is added
+        group_data = self._read_job_group_from_disk()
+
+        saved_job_info = group_data['job_group_data']
+        saved_job_info.append(updated_info)
+
+        # self._group_info['job_group_data'] = saved_job_info  # save changes in object
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:186]: Don't we want what this line do ?

File: perceval/runtime/job_group.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        group_data = self._read_job_group_from_disk()
+
+        self._group_info['created_date'] = group_data['created_date']
+        self._group_info['modified_date'] = JobGroup._get_current_datetime()
+        self._group_info['job_group_data'] = group_data['job_group_data']
+        # Write to disk (necessary to update modification time)
+        self._write_job_group_to_disk()
+
+    @staticmethod
+    def list_saved_job_groups():
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:133]: missing typehinting for return object

File: perceval/runtime/job_group.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        List of necessary information (id,status,metadata) for each job within the group
+        """
+        return self._group_info['job_group_data']
+
+    @staticmethod
+    def _get_current_datetime():
+        return datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+
+    @staticmethod
+    def _get_job_group_dir():
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:81]: missing typehinting for return object

File: perceval/runtime/job_group.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+    @property
+    def job_group_data(self):
+        """
+        List of necessary information (id,status,metadata) for each job within the group
+        """
+        return self._group_info['job_group_data']
+
+    @staticmethod
+    def _get_current_datetime():
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:77]: missing typehinting for return object

File: perceval/runtime/job_group.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+    @property
+    def name(self):
+        """
+        Name of the job Group [also the filename (without extension) on disk]
+        """
+        return self._name
+
+    @property
+    def job_group_data(self):
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:70]: missing typehinting for return object

File: perceval/runtime/job_group.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+
+        if self._job_group_exists(name):
+            get_logger().info(f'Job Group with name {name} exists, loading from disk', channel.general)
+            self._load_job_group()
+        else:
+            self._create_job_group()
+
+
+    @property
+    def name(self):
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:63]: missing typehinting for return object

File: perceval/runtime/job_group.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+    def list_saved_job_groups():
+        """
+        Returns a list of filenames of all JobGroups saved to disk
+        """
+        jgrp_path = JobGroup._get_job_group_dir()
+        files = [f for f in os.listdir(jgrp_path) if f.endswith(FILE_EXT_JGRP)]
+        return files
+
+    @staticmethod
+    def _job_group_exists(name: str):
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:142]: missing typehinting for return object

File: perceval/runtime/job_group.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+        # modifies the recorded information when a new job is added
+        group_data = self._read_job_group_from_disk()
+
+        saved_job_info = group_data['job_group_data']
+        saved_job_info.append(updated_info)
+
+        # self._group_info['job_group_data'] = saved_job_info  # save changes in object
+        self._write_job_group_to_disk()  # save changes to disk
+
+    def progress(self):
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:189]: missing typehinting for return object

File: perceval/runtime/job_group.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+# exqalibur is limited to the python bindings used by Perceval.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import warnings
```
Comment (MarionQuandela) [inline perceval/runtime/job_group.py:29]: Do not use the warnings module, instead use `get_logger().warn(..., channel.user)`

**Review by MarionQuandela — CHANGES_REQUESTED** [review]

File: perceval/runtime/job_group.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+    @staticmethod
+    def _get_current_datetime():
+        return datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+
+    @staticmethod
+    def _get_job_group_dir():
+        return os.path.join(PersistentData().directory, "job_group")
+
+    @staticmethod
+    def _setup_job_group_dir(dir_path):
```
[the remark this answers is not in the document]
Comment (ericbrts) [reply perceval/runtime/job_group.py:85]: Same if `_get_job_group_dir` turns into a class variable, as suggested by Marion

File: perceval/runtime/job_group.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+    """
+    A JobGroup handles a collection of Jobs saved on disk using perceval
+    persistent data. It can perform various tasks such as
+    - Saving information for a collection of jobs, whether they have been sent to the cloud or not.
+    - Running jobs within the group either in parallel or sequentially.
+    - Rerunning failed jobs within the group.
+    """
+    def __init__(self, name: str):
+        self._name = name
+        self._group_info = dict()
```
Comment (ericbrts) [inline perceval/runtime/job_group.py:52]: It's better to store a `PersistentData` instance in your class (it'll help doing the mocks for unit tests at least) Check how it's done in `Config` or `TokenProvider` classes for reference.

File: perceval/runtime/job_group.py [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+        return datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+
+    @staticmethod
+    def _get_job_group_dir():
+        return os.path.join(PersistentData().directory, "job_group")
+
+    @staticmethod
+    def _setup_job_group_dir(dir_path):
+        # creates a sub folder in persistent data directory if non-existent
+        try:
```
Comment (ericbrts) [inline perceval/runtime/job_group.py:87]: This code should probably be added to `PersistentData`. I'd suggest a `create_sub_directory(relative_path)` method that you'd call with a path relative to the root of persistent data.
In your case:  `self._persistent_data.create_sub_directory("job_group")`

File: perceval/runtime/job_group.py (hunk 3 above)
Comment (ericbrts) [reply perceval/runtime/job_group.py:133]: Also, the method name is a bit long, and is called by typing the class name anyway. So I'd suggest:
`JobGoup.list_existing()`

File: perceval/runtime/job_group.py [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+        self._write_job_group_to_disk()
+
+    @staticmethod
+    def list_saved_job_groups():
+        """
+        Returns a list of filenames of all JobGroups saved to disk
+        """
+        jgrp_path = JobGroup._get_job_group_dir()
+        files = [f for f in os.listdir(jgrp_path) if f.endswith(FILE_EXT_JGRP)]
+        return files
```
[the remark this answers is not in the document]
Comment (ericbrts) [reply perceval/runtime/job_group.py:139]: I guess we want to get rid of the file extension here.

File: perceval/runtime/job_group.py [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+        else:
+            job_info['status'] = job_to_add.status()
+
+        if not job_info['status'] == 'SUCCESS':
+            # save metadata for any job that did not run successfully or was not sent
+            job_info['metadata'] = {'headers': job_to_add._rpc_handler.headers,
+                                    'platform': job_to_add._rpc_handler.name,
+                                    'url': job_to_add._rpc_handler.url}
+
+        self._group_info['job_group_data'].append(job_info)  # save changes in object
```
Comment (ericbrts) [inline perceval/runtime/job_group.py:175]: Do you have a system refusing job id duplicates?

File: perceval/runtime/job_group.py [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+                    # remove metadata information if job ran successfully
+                    del job_entry['metadata']
+                job_entry['status'] = rj_status  # update current status
+
+            status_jobs_in_group.append(job_entry['status'])
+
+        self._write_job_group_to_disk()  # rewrites the group information on disk
+
+        # replace None with 'Not_sent' - more readable
+        status_list = [status if status is not None else "Not_sent" for status in status_jobs_in_group]
```
Comment (ericbrts) [inline perceval/runtime/job_group.py:211]: What are users going to do with a list of statuses? We can answer this question in the other ticket, but right now I guess the progress() return value isn't really usable.

File: perceval/runtime/job_group.py [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+                                 platform_metadata['url'], user_token)
+
+        return RemoteJob.from_id(job_entry['id'], rpc_handler)
+
+    def run_parallel(self, **kwargs):
+        # todo: refine and implement
+        pass
+
+    def run_sequential(self, delay: float, **kwargs):
+        # todo : refine and implement
```
[the remark this answers is not in the document]
Comment (ericbrts) [reply perceval/runtime/job_group.py:229]: I'd even say: don't write the code at all if it's not meant to be. You can add these methods later on.

File: perceval/runtime/job_group.py [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+        To delete all existing Job groups on disk
+        """
+        jgrp_dir_path = JobGroup._get_job_group_dir()
+        list_groups = JobGroup.list_saved_job_groups()
+        ps = PersistentData()
+        for each_file in list_groups:
+            ps.delete_file(os.path.join(jgrp_dir_path, each_file))
+
+    @staticmethod
+    def delete_job_group(filename: str):
```
Comment (ericbrts) [inline perceval/runtime/job_group.py:251]: Again, you shouldn't expect users to know about files and especially file extensions. The signature here should just take a job group name, not a filename.

File: perceval/runtime/job_group.py [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+        """
+        jgrp_dir_path = JobGroup._get_job_group_dir()
+
+        PersistentData().delete_file(os.path.join(jgrp_dir_path, filename))
+
+    @staticmethod
+    def delete_job_groups_date(del_before_date: int):
+        """
+        Delete all saved Job Groups created before a date (not included).
+        :param del_before_date: integer (form - YYYYMMDD) files created before this date deleted
```
[the remark this answers is not in the document]
Comment (ericbrts) [reply perceval/runtime/job_group.py:264]: Oh yes, please don't invent a new way to encode a date :D

**Review by ericbrts — CHANGES_REQUESTED** [review]

File: perceval/runtime/job_group.py [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+        # validate read/write permissions for the directory
+        if not os.access(dir_path, os.W_OK) or not os.access(dir_path, os.R_OK):
+            warnings.warn(UserWarning(f"Cannot read or write in {dir_path}"))
+
+    def _create_job_group(self):
+        # Saves information for a new job group on disk using PersistentData(),
+        jgrp_dir_path = JobGroup._get_job_group_dir()
+        JobGroup._setup_job_group_dir(jgrp_dir_path)  # create directory and validat permissions
+
+        self._group_info['created_date'] = JobGroup._get_current_datetime()
```
[the remark this answers is not in the document]
Comment (raksharuia) [reply perceval/runtime/job_group.py:102]: I did test that they were the same after running many tests. However, I will update the code to include your suggestion to be on the safer side.

File: perceval/runtime/job_group.py [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+        self._group_info['modified_date'] = JobGroup._get_current_datetime()
+        self._group_info['job_group_data'] = []
+        # write to disk
+        self._write_job_group_to_disk()
+
+    def _read_job_group_from_disk(self) -> dict:
+        # returns the Job group data stored on disk
```
[the remark this answers is not in the document]
Comment (raksharuia) [reply perceval/runtime/job_group.py:109]: Oops my bad, fixed all of them.

File: perceval/runtime/job_group.py [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+        list_groups = JobGroup.list_saved_job_groups()
+        ps = PersistentData()
+        for each_file in list_groups:
+            ps.delete_file(os.path.join(jgrp_dir_path, each_file))
+
+    @staticmethod
+    def delete_job_group(filename: str):
+        """
+        Delete a single JobGroup file by its name
+        :param filename: a JobGroup filename with its extension to delete
```
[the remark this answers is not in the document]
Comment (raksharuia) [reply perceval/runtime/job_group.py:254]: already there.

File: perceval/runtime/job_group.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (raksharuia) [reply perceval/runtime/job_group.py:211]: my bad, thanks for noticing this

File: perceval/runtime/job_group.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (raksharuia) [reply perceval/runtime/job_group.py:85]: changed dir_path to class variable and updated usage.

File: perceval/runtime/job_group.py (hunk 2 above)
Comment (raksharuia) [reply perceval/runtime/job_group.py:186]: I had modified how the data is modified between `add` and `modify` methods. Forgot to remove this line.

File: perceval/runtime/job_group.py (hunk 16 above)
Comment (raksharuia) [reply perceval/runtime/job_group.py:211]: Yes, in this state it is not usable. The idea of having this method here is simply to have the structure of the method ready, ii.e. how it behaves for each job in the group. Do you want me to update it to not return anything here? Or is it ok if we leave it and finish the `progress` method in the next ticket?

File: perceval/runtime/job_group.py (hunk 15 above)
Comment (raksharuia) [reply perceval/runtime/job_group.py:175]: I did not, implemented.

**Review by ericbrts — APPROVED** [review]

File: perceval/runtime/job_group.py [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+        jgrp_dir_path = JobGroup._JGRP_DIR_PATH
+        list_groups = JobGroup.list_existing()
+        for each_file in list_groups:
+            JobGroup._JGRP_PERSISTENT_DATA.delete_file(os.path.join(jgrp_dir_path, each_file))
+
+    @staticmethod
+    def delete_job_group(filename: str):
+        """
+        Delete a single JobGroup file by its name
+        :param filename: a JobGroup name to delete
```
[the remark this answers is not in the document]
Comment (raksharuia) [reply perceval/runtime/job_group.py:256]: Oh yes, thanks for noticing this :-)

File: perceval/runtime/job_group.py [hunk 24]
```diff
@@ -0,0 +1,281 @@
...
+
+    @staticmethod
+    def delete_all_job_groups():
+        """
+        To delete all existing Job groups on disk
+        """
+        jgrp_dir_path = JobGroup._JGRP_DIR_PATH
+        list_groups = JobGroup.list_existing()
+        for each_file in list_groups:
+            JobGroup._JGRP_PERSISTENT_DATA.delete_file(os.path.join(jgrp_dir_path, each_file))
```
[the remark this answers is not in the document]
Comment (raksharuia) [reply perceval/runtime/job_group.py:250]: Oops

**Pull request merged**